Source/Core/Core/IOS/Network/KD/VFF/VFFUtil.cpp

```diff
@@ -0,0 +1,309 @@
+// Copyright 2022 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "Core/IOS/Network/KD/VFF/VFFUtil.h"
+
+#include <algorithm>
+#include <cmath>
+#include <vector>
+
+#include <fmt/format.h>
+
+// Does not compile if diskio.h is included first.
+// clang-format off
+#include "ff.h"
+#include "diskio.h"
+// clang-format on
+
+#include "Common/Align.h"
+#include "Common/FatFsUtil.h"
+#include "Common/Logging/Log.h"
+#include "Common/ScopeGuard.h"
+#include "Common/Swap.h"
+
+#include "Core/IOS/Uids.h"
+
+static DRESULT read_vff_header(IOS::HLE::FS::FileHandle* vff, FATFS* fs)
+{
+  struct IOS::HLE::NWC24::VFFHeader header;
+  if (!vff->Read(&header, 1))
+  {
+    ERROR_LOG_FMT(IOS_WC24, "Failed to read VFF header.");
+    return RES_ERROR;
+  }
+
+  u16 cluster_size = 0;
+  u16 cluster_count = 0;
+
+  switch (Common::swap16(header.endianness))
+  {
+  case IOS::HLE::NWC24::VF_BIG_ENDIAN:
+    cluster_size = Common::swap16(header.clusterSize) * 16;
+    cluster_count = Common::swap32(header.volumeSize) / cluster_size;
+    break;
+  case IOS::HLE::NWC24::VF_LITTLE_ENDIAN:
+    // TODO: Actually implement.
+    // RiiConnect24's VFF are the only seen instance of Little Endian as it uses an SDK tool.
+    // Another option is to just delete these VFF's and let the current channel create a Big Endian
+    // one.
+    return RES_ERROR;
+  }
+
+  if (cluster_count < 4085)
+  {
+    fs->fs_type = FS_FAT12;
+
+    u32 table_size = std::floor((cluster_count + 1) / 2) * 3;
+    table_size = (table_size + cluster_size - 1) & ~(cluster_size - 1);
```

Comment:
Not critical, but since your friend pointed me at this line in IRC: This should be equivalent to:

```
Common::AlignUp(table_size, cluster_size)
```
(from `Common/Align.h`)

Same in the other branch below.

I'm pretty sure that `std::floor` in the line above it also does nothing since it's integers so it can just be `u32 table_size = ((cluster_count + 1) / 2) * 3;`.